File: 00070-climbing-stairs/main2.py
dp = [-1 for _ in range(46)]
dp[0] = 1
dp[1] = 1


def aux(n):
    if dp[n] == -1:
        dp[n] = aux(n - 1) + aux(n - 2)
    return dp[n]


class Solution:
    def climbStairs(self, n: int) -> int:
        return aux(n)


sol = Solution()
ans = sol.climbStairs(11)
print("ans", ans)

File: 377-combination-sum-iv/main2.cpp
#include <bits/stdc++.h>
#include "/home/ealjkl/competitive-programming/utils/cpp/print.cpp"

using namespace std;

map<int, int> cache;

int cachedFunc(vector<int> &nums, int target, map<int, int> &cache) {
    int ans = 0;
    if (target < 0) {
        return 0;
    }
    if (target == 0) {
        return 1;
    }
    if (cache.count(target)) {
        return cache[target];
    }

    for (auto num: nums) {
        int combinations = cachedFunc(nums, target - num, cache);
        ans += combinations;
    }
    cache[target] = ans;
    return ans;
}

class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        cache = {};
        return cachedFunc(nums, target, cache);
    }
};

int main() {
    Solution sol {};
    // vector<int> x = {1, 2, 3};
    vector<int> y = {3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25};
    // cout << sol.combinationSum4(x, 4) << endl;
    cout << sol.combinationSum4(y, 10) << endl;
}